File: src/vcmp/functions/keybind.py
from typing import NamedTuple, Optional, overload, final
from vcmp.__export import funcs
from vcmp.types import KeyCode

class KeyBind(
    NamedTuple
):
    slot: int
    can_release: bool
    key1: KeyCode
    key2: Optional[KeyCode]
    key3: Optional[KeyCode]

    def __hash__(self) -> int:
        return self.slot

@overload
def bindkey(
    can_release: bool,
    key: KeyCode,
) -> KeyBind:
    ...

@overload
def bindkey(
    can_release: bool,
    key: KeyCode,
    key2: KeyCode,
) -> KeyBind:
    ...

def bindkey(
    can_release: bool,
    key: KeyCode,
    key2: Optional[KeyCode] = None,
    key3: Optional[KeyCode] = None,
) -> KeyBind:
    """Bind a key to trigger event call."""
    keys = [
        key,
        key2 if key2 is not None else 0,
        key3 if key3 is not None else 0,
    ]
    slot = funcs.get_key_bind_unused_slot()
    funcs.register_key_bind(slot, can_release, *keys)

    data = funcs.get_key_bind_data(slot)
    return KeyBind(
        slot,
        data[0],
        data[1],
        data[2] if data[2] != -1 else None,
        data[3] if data[3] != -1 else None,
    )

def get_bindkey(
    slot: int
):
    """Get binded key from slot."""
    try:
        data = funcs.get_key_bind_data(slot)
        return KeyBind(
            slot,
            data[0],
            data[1],
            data[2] if data[2] != -1 else None,
            data[3] if data[3] != -1 else None,
        )
    except:
        return None
    
def remove_bindkey(
    slot: int
):
    """
        Remove binded key from slot.
    """
    try:
        funcs.remove_key_bind(slot)
        return True
    except:
        return False
    
def remove_all_bindkeys():
    """Remove all binded keys."""
    funcs.remove_all_key_binds()
    

File: crates/vcmp_bindings/src/func/keybind.rs
use crate::func::VcmpFunctions;

pub trait QueryKeyBind {
}
pub trait SetKeyBind {
}

impl SetKeyBind for VcmpFunctions {
    fn register_key_bind(
        &self,
        release: bool,
        key1: i32,
        key2: Option<i32>,
        key3: Option<i32>,
    ) -> Keybind {
        let k1 = key1;
        let k2 = key2.unwrap_or(0);
        let k3 = key3.unwrap_or(0);
        (self.inner.RegisterKeyBind)(
            slot,
            release as u8,
            k1,
            k2,
            k3,
        );

        Keybind { slot, can_release: release, key1: k1, key2: k2, key3: k3 }
    }
    fn remove_key_bind(&self, slot: i32) {
        (self.inner.RemoveKeyBind)(slot);
    }
    fn remove_all_key_binds(&self) {
        (self.inner.RemoveAllKeyBinds)();
    }
}

impl QueryKeyBind for VcmpFunctions {
    fn get_key_bind_unused_slot(&self) -> i32 {
        (self.inner.GetKeyBindUnusedSlot)()
    }
    fn get_key_bind_data(&self, slot: i32) -> Keybind {
        let (mut release, mut key1, mut key2, mut key3) = (0_u8, 0, 0, 0);
        (self.inner.GetKeyBindData)(slot, &mut release, &mut key1, &mut key2, &mut key3);
        
        Keybind { slot, can_release: release != 0, key1, key2, key3 }
    }
}
